### PR TITLE
Add surviving of the project

### LINKED ISSUE #734: Projects should survive integrity break on load
Now if we’d try to open a project which is fine from JSON and data shape point of view, but which have integrity issues such as bad patch path references and bad link references, we’ll fail. IDE hangs. That’s very frustrating. We need a mechanism which allows to go on.

We can still draw broken nodes and links with some special style (dashed red). There should be enough information for that: patch path is known, the fact the pin was there is known too. The info about the label is lost, but that’s OK.

### Test scenario

1. Make a blink patch: clock → flip-flop → led
2. Save it and close IDE
3. Delete the `flip-flop` from `lib/xod/core`
4. Launch IDE and try to open that patch

### Acceptance criteria

- [x] The patch gets opened successfully. The `flip-flop` and its links are drawn with the broken style
- [x] I can select the broken node, drag it around, select, delete
- [x] I can select the broken links and delete them
- [x] I can save/load a project with broken stuff and get the same result
- [x] When I try to transpile I get a clear error message like “Patch @/main contains broken references. Fix or delete them to continue.”
- [x] I can close IDE, restore the `flip-flop`, open the project again and get the state I’d get if the `flip-flop` was never deleted
- [x] The same round trip is possible for another scenario when I only delete the `TGL` pin in `flip-flop`. In that case, only a single link would be considered broken.

### PR DESCRIPTION
**Pull request by brusherru**
It closes #734 .

### CODE REVIEW

File: packages/xod-client/src/project/selectors.js [hunk 1]
```diff
@@ -180,13 +172,28 @@ export const getRenderableLinks = createMemoizedSelector(
...
         const outputPinKey = XP.getLinkOutputPinKey(link);
 
+        const inputNodeIsDead = nodes[inputNodeId].dead;
+        const outputNodeIsDead = nodes[outputNodeId].dead;
+        const inputPinKeyHasDeadType = (
+          nodes[inputNodeId].pins[inputPinKey].type === XP.PIN_TYPE.DEAD
+        );
+        const outputPinKeyHasDeadType = (
+          nodes[outputNodeId].pins[outputPinKey].type === XP.PIN_TYPE.DEAD
+        );
```
Comment (nkrkv) [inline packages/xod-client/src/project/selectors.js]: `R.pathEq` would be more compact and safer.

File: packages/xod-project/src/patch.js [hunk 2]
```diff
@@ -887,3 +887,123 @@ export const removeDebugNodes = def(
...
   )(patch)
 );
+
+/**
+ * Returns a map of pins for patch that exists.
+ * So Pins are fully valid, contains proper types and etc.
+ */
+export const getPinsForExistentPatch = def(
+  'getPinsForExistentPatch :: Node -> Patch -> Map PinKey Pin',
+  (node, patch) => R.compose(
```
Comment (nkrkv) [inline packages/xod-project/src/patch.js]: As far as I understand, we’re getting pins for a node, not for a patch. So the name is misguiding. IMO better: `listExistingNodePins`

Given that, the function is better suitable for `node.js`. In that case, it hasn't even to be exported.

Also, BTW, “existent” is awkward, “existing“ is much more casual.

File: packages/xod-project/src/patch.js [hunk 3]
```diff
@@ -887,3 +887,123 @@ export const removeDebugNodes = def(
...
+  )(patch)
+);
+
+/**
+ * Returns a map of pins for Node, that points to unexistent Patch.
+ * These pins are created from links, connected with some Pins in this Node.
+ */
+export const getPinsForUnexistentPatch = def(
+  'getPinsForUnexistentPatch :: Node -> Patch -> Map PinKey Pin',
+  (node, currentPatch) => {
```
Comment (nkrkv) [inline packages/xod-project/src/patch.js]: The same as above: `listDeadNodePins`.

Not obvious why the second arg is called “current” patch.

BTW, “unexsistent” is exotic… “nonexisting” is casual.

File: packages/xod-project/src/patch.js [hunk 4]
```diff
@@ -887,3 +887,123 @@ export const removeDebugNodes = def(
...
+        R.always(CONST.PIN_DIRECTION.OUTPUT),
+      )),
+      listLinksByNode(node)
+    )(currentPatch);
+  }
+);
+
+/**
+ * Some Patch could have a dead link, and for this case we should
+ * add "unknown" pin to the Patch, to connect this link somewhere.
```
Comment (nkrkv) [inline packages/xod-project/src/patch.js]: *dead* pin

File: packages/xod-project/src/patch.js [hunk 5]
```diff
@@ -887,3 +887,123 @@ export const removeDebugNodes = def(
...
+ * E.G.
+ * There was a link to the `xod/core/flip-flop` pin `TGL`. But then
+ * Pin `TGL` was destroyed somehow and left only `SET` and `RST`. So
+ * we have to add `TGL` pin to current `xod/core/flip-flop` with
+ * unknown type, to render link somehow and mark it as dead.
```
Comment (nkrkv) [inline packages/xod-project/src/patch.js]: dead type

File: packages/xod-project/src/patch.js [hunk 6]
```diff
@@ -887,3 +887,123 @@ export const removeDebugNodes = def(
...
+  'ensureLinksConnectedToPins :: Node -> Patch -> Map PinKey Pin -> Map PinKey Pin',
+  (node, currentPatch, pins) => {
+    const nodeId = Node.getNodeId(node);
+    const pinKeys = R.keys(pins);
+    const pinsByDir = R.applySpec({
+      [CONST.PIN_DIRECTION.INPUT]: R.filter(Pin.isInputPin),
+      [CONST.PIN_DIRECTION.OUTPUT]: R.filter(Pin.isOutputPin),
+    })(R.values(pins));
+
+    const rejectLinkConnectedWithExistentPins = R.reject(
```
Comment (nkrkv) [inline packages/xod-project/src/patch.js]: existent → existing

File: packages/xod-project/src/patch.js [hunk 7]
```diff
@@ -887,3 +887,123 @@ export const removeDebugNodes = def(
...
+ * add "unknown" pin to the Patch, to connect this link somewhere.
+ * E.G.
+ * There was a link to the `xod/core/flip-flop` pin `TGL`. But then
+ * Pin `TGL` was destroyed somehow and left only `SET` and `RST`. So
+ * we have to add `TGL` pin to current `xod/core/flip-flop` with
+ * unknown type, to render link somehow and mark it as dead.
+ */
+export const ensureLinksConnectedToPins = def(
+  'ensureLinksConnectedToPins :: Node -> Patch -> Map PinKey Pin -> Map PinKey Pin',
+  (node, currentPatch, pins) => {
```
Comment (nkrkv) [inline packages/xod-project/src/patch.js]: ~current~Patch

Would a name `upsertDeadPins` fit better?

File: packages/xod-project/src/pin.js [hunk 8]
```diff
@@ -35,6 +35,23 @@ export const createPin = def(
   })
 );
 
+/**
+ * @private
+ */
+export const createUnknownPin = def(
```
Comment (nkrkv) [inline packages/xod-project/src/pin.js]: `createDeadPin`

File: packages/xod-project/test/project.spec.js [hunk 9]
```diff
@@ -670,4 +674,30 @@ describe('Project', () => {
...
+      Patch.getNodeById
+    );
+
+    it('should add broken pins to nodes, that points to unexisting patch', () => {
+      const pinKeysOut = getPinKeysByNodeId('brokenNodeOutLinks', curPatch);
+      assert.lengthOf(pinKeysOut, 1);
+      const pinKeysIn = getPinKeysByNodeId('brokenNodeInLinks', curPatch);
+      assert.lengthOf(pinKeysIn, 2);
+    });
+    it('should add broken pins to valid nodes, that has connected links to unexisting pins', () => {
```
Comment (nkrkv) [inline packages/xod-project/test/project.spec.js]: unexisting → non-existing

File: packages/xod-project/test/project.spec.js [hunk 10]
```diff
@@ -670,4 +674,30 @@ describe('Project', () => {
...
+    const getPinKeysByNodeId = R.compose(
+      Maybe.maybe([], R.identity),
+      R.map(R.compose(
+        R.keys,
+        Project.getPinsForNode(R.__, curPatch, project)
+      )),
+      Patch.getNodeById
+    );
+
+    it('should add broken pins to nodes, that points to unexisting patch', () => {
```
Comment (nkrkv) [inline packages/xod-project/test/project.spec.js]: non-existing

File: packages/xod-arduino/src/transpiler.js [hunk 11]
```diff
@@ -390,11 +391,11 @@ const transformProjectWithImpls = def(
...
     )),
     R.chain(Project.flatten(R.__, path, impls)),
-    R.ifElse(
+    R.when(
       () => (opts.debug === true),
-      Either.of,
-      Project.updatePatch(path, Project.removeDebugNodes)
-    )
+      R.chain(Project.updatePatch(path, Project.removeDebugNodes))
+    ),
```
Comment (nkrkv) [inline packages/xod-arduino/src/transpiler.js]: O_o. When debug, remove debug nodes? I think an exact opposite of it is required. Would be simpler to just:

```javascript
proj => opts.debug ? R.chain(Project.updatePatch(path, Project.removeDebugNodes), proj) : proj;
```

Please, verify debug mode works correctly after the change.

File: packages/xod-project/src/constants.js [hunk 12]
```diff
@@ -6,6 +6,8 @@
  * @enum {string}
  */
 export const ERROR = {
+  // dead references
+  DEAD_REFERENCES_FOUND: 'Patch "{patchPath}" contains dead references:\n{message}\nFix or delete them to continue.',
```
Comment (nkrkv) [inline packages/xod-project/src/constants.js]: That’s what I talked about. Not i18n friendly. Should be:

```javascript
export const formatDeadReferencesFound = (patchPath, submessage) => [
  `Patch ${patchPath} contains dead references:`
  submessage,
  'Fix or delete them to continue.',
].join('\n');
```

It’s not a constant. So it should live in `messages.js`. I suggest moving it there and leave a `// TODO:` about legacy for existing “constants”.

File: packages/xod-project/src/utils.js [hunk 13]
```diff
@@ -31,6 +31,30 @@ export const formatString = R.curry((template, replacements) =>
 );
 
 /**
+ * Updates Error message of Left value by wrapping it with
+ * dead reference error message. Right value will be passed unchanged.
+ *
+ * Error message updated by mutation (to prevent creating new stack trace).
```
Comment (nkrkv) [inline packages/xod-project/src/utils.js]: I suggest moving this note closer to `// eslint-disable-next-line no-param-reassign`

**Comment by brusherru** [comment]
@nkrkv fixed!

File: packages/xod-project/src/patch.js [hunk 14]
```diff
@@ -887,3 +887,123 @@ export const removeDebugNodes = def(
...
+        R.compose(isAmong(pinKeys), Link.getLinkOutputPinKey),
+      )
+    );
+
+    return R.compose(
+      R.merge(pins),
+      R.indexBy(Pin.getPinKey),
+      R.when(
+        R.complement(R.isEmpty),
+        R.compose(
```
Comment (evgenykochetkov) [inline packages/xod-project/src/patch.js]: This composition is a bit long and confusing.
I'd love to see the types of some of the intermediate results.

File: packages/xod-project/src/project.js [hunk 15]
```diff
@@ -122,15 +122,6 @@ export const setProjectLicense = def(
   R.assoc('license')
 );
 
-// TODO: remove
```
Comment (evgenykochetkov) [inline packages/xod-project/src/project.js]: 👍🏻

File: packages/xod-project/src/utils.js [hunk 16]
```diff
@@ -31,6 +32,26 @@ export const formatString = R.curry((template, replacements) =>
...
 /**
+ * Updates Error message of Left value by wrapping it with
+ * dead reference error message. Right value will be passed unchanged.
+ *
+ * @param {PatchPath} patchPath
+ * @param {Either} either Either Error a
+ * @returns {Either} Either Error a
+ */
+export const wrapDeadRefErrorMessage = R.curry(
+  (patchPath, either) => either.bimap(
```
Comment (evgenykochetkov) [inline packages/xod-project/src/utils.js]: Yay! Bifunctors!

**Pull request merged**